### ISSUE TITLE
Improve IFlxEffects

### ISSUE THREAD

**Issue by adrianulima** [HaxeFlixel/flixel-addons#228]
This issue is to list next improvements on `IFlxEffect` that is planned, and if someone have a new idea for it:

FlxTrailEffect:
- [ ] Add support for angle/rotation;
- [ ] Add support for flip;
- [ ] Add support for scale;

FlxOutlineEffect:
- [ ] Add property `quality` similar to `FlxText#borderQuality`;
- [ ] Add `FlxOutlineMode` enum;
- [ ] Add `FlxOutlineMode.NORMAL` similar to `FlxTextBorderStyle.OUTLINE`;
- [ ] Add `FlxOutlineMode.FAST` similar to `FlxTextBorderStyle.OUTLINE_FAST`;

New IFlxEffects:
- [ ] Add `FlxShadowEffect` similar to `FlxTextBorderStyle.SHADOW`;

General:
- [ ] Try to reduce `BitmapData` new instances.

I think `IFlxEffect` is too young yet and have things to be improved, but when people start using it and help us it will be very important to many other cases.
`FlxCamera` has it own `shake` method that can be replaced by `FlxShakeEffect`;
`FlxText` has it own `border` that can be replaced by `FlxOutlineEffect` + `FlxShadowEffect`.

**Comment by Gama11** [comment]
I'm not sure `FlxText` border styles should be replaced by `FlxEffectSprite` effects... You'd need an additional `FlxEffectSprite` instance just to get an outline effect.

**Comment by adrianulima** [comment]
Yeah, what I'm saying is that `IFlxEffect` has potential to be more used in the future when we find all problems, implement all improvements and add more effects. Probably a far future `(5.0.0)`.

I thought about instead of a class`FlxEffectSprite` if we could implement all the logic in the bitmapData: `FlxBitmapData`.

**Comment by xerosugar** [comment]
Suggestion for a new type of effect (or an "upgraded" version of FlxWaveEffect maybe?)
It works almost like the wave effect we have now, but the one I'm thinking adds distortion in a different way. In old 2D games the effect was often used in underwater scenes or to give the impression of heat (like in deserts).

Examples:
https://youtu.be/rXKyFZ-O6yE?t=51 (under water)
https://youtu.be/X8sXXNUc9GI?t=265 (under water, a bit more discrete)
https://youtu.be/69xfXl3rBoM?t=7 (desert)

I've looked around for some snippets and this is what I've found that seems usable (they distort both horizontally and vertically, but the above examples only do it horizontally)

- http://ariya.ofilabs.com/2012/03/underwater-effect-with-html5-canvas.html
(here's the whole js-file: https://raw.githubusercontent.com/ariya/ariya.github.com/master/canvas/underwater/underwater.js)

- http://wonderfl.net/c/hM3G

**Comment by xerosugar** [comment]
Also, a suggestion for the FlxMosaic effect I made a demo for:
I'd be great if you could tell FlxMosaic what to run the effect on (instead of providing it with an image asset), like:
a) pixelate everything that's added *before* it in the state but not what's added after it, and
b) pixelate the whole picture

**Comment by adrianulima** [comment]
Wow nice effect and cool examples / snippets, this would be awesome. I think it works different to `FlxWaveEffect` enough to be a new class. Maybe named `FlxRippleEffect` or `FlxDistortionEffect`, @Gama11 is better naming new features.

I will add it to the list, but I don't know when I will have free time to implement it.
